### PR TITLE
add look-back to service-now fetch incidents

### PR DESCRIPTION
**Pull request by GuyAfik**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/48145

## Description
Introduce the look-back for the service now fetch incidents.

## Screenshots
Paste here any images that will help the reviewer

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2.py [hunk 1]
```diff
@@ -1992,93 +1994,130 @@ def document_route_to_table(client: Client, args: dict) -> Tuple[Any, Dict[Any,
...
 
-        incidents.append({
-            'name': f"ServiceNow Incident {result.get(client.incident_name)}",
-            'labels': labels,
-            'details': json.dumps(result),
-            'severity': severity,
-            'attachment': file_names,
-            'rawJSON': json.dumps(result)
-        })
+    demisto.info(f'last run at the end of the incidents fetching {last_run}')
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2.py:2114]:
```suggestion
    demisto.debug(f'last run at the end of the incidents fetching {last_run}')
```

File: Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py [hunk 2]
```diff
@@ -1,3 +1,6 @@
+from CommonServerPython import timedelta
+from datetime import datetime
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py:2]:
```suggestion
from datetime import datetime, timedelta
```

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py [hunk 3]
```diff
@@ -334,14 +337,298 @@ def test_fetch_incidents_with_incident_name(mocker):
...
     incidents = fetch_incidents(client)
     assert incidents[0].get('name') == 'ServiceNow Incident Unable to access Oregon mail server. Is it down?'
 
 
+class TestFetchIncidentsWithLookBack:
+    LAST_RUN = {}
+
+    API_TIME_FORMAT = '%Y-%m-%d %H:%M:%S'
+
+    NOW = datetime.utcnow()
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py:353]: better not to use real time, such tests tend to be flaky. consider using freezegun

File: Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py [hunk 4]
```diff
@@ -3002,7 +3007,9 @@
                 'cmdb_ci': {'link': 'demisto.com', 'value': 'test'}, 'delivery_plan': '', 'impact': '2',
                 'active': 'true',
                 'work_notes_list': '', 'business_service': '', 'priority': '4', 'sys_domain_path': '/', 'rfc': '',
-                'time_worked': '', 'expected_start': '', 'opened_at': '2019-09-05 00:41:01', 'business_duration': '',
+                'time_worked': '', 'expected_start': '',
+                'opened_at': (datetime.utcnow() - timedelta(minutes=15)).strftime('%Y-%m-%d %H:%M:%S'),
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py:3011]: same here

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py [hunk 5]
```diff
@@ -334,14 +337,298 @@ def test_fetch_incidents_with_incident_name(mocker):
...
+        - fetch incidents parameters including look back according to their opened time.
+
+        When
+        - trying to fetch incidents for 3 rounds.
+
+        Then
+        - first fetch - should fetch incidents 2, 4, 5 (because only them match the query)
+        - second fetch - should fetch incident 3 (because now incident 2, 4, 5, 3 matches the query too)
+        - third fetch - should fetch incident 1 (because now incident 2, 4, 5, 3, 1 matches the query too)
+        - forth fetch - should fetch nothing as there are not new incidents who match the query
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py:442]:
```suggestion
        - fourth fetch - should fetch nothing as there are not new incidents who match the query
```

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py [hunk 6]
```diff
@@ -334,14 +337,298 @@ def test_fetch_incidents_with_incident_name(mocker):
...
+    API_TIME_FORMAT = '%Y-%m-%d %H:%M:%S'
+
+    NOW = datetime.utcnow()
+    # used for fetching tickets without look back
+
+    def set_last_run(self, new_last_run):
+        self.LAST_RUN = new_last_run
+
+    @pytest.mark.parametrize(
+        'start_incidents, phase2_incident, phase3_incident, look_back',
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py:360]: I see there are two cases here, please describe the differences between them in the docstring

File: Packs/ServiceNow/Integrations/ServiceNowv2/README.md [hunk 7]
```diff
@@ -106,6 +107,11 @@ which you define in the instance configuration: ticket_type, query, and limit.
 For the first fetch, the integration will fetch incidents that were created 10 minutes earlier. 
 After that, the integration will fetch incidents created after the timestamp of the last fetch.
 
+Note:
+In case the **lookback** parameter is initialized with '0' and during that time incidents were fetched, if changing 
+the look back to a number that is greater > 0, then in the initial fetching there could be incidents duplications.
+If the integration was already set with look back > 0 in the first place, then those incident duplications would not occur.
```
Comment (dantavori) [inline Packs/ServiceNow/Integrations/ServiceNowv2/README.md:113]: basically increasing the value of the parameter, even from 10 to 20 for example, will cause the same issue. let's explain this, and that it's by design but it's important to be aware.
please add this also in the `_description.md`.

**Review by dantavori — CHANGES_REQUESTED** [review]
looks great, you can already send the yml to the customer and fix my CR notes later.

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py (hunk 3 above)
Comment (GuyAfik) [reply Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py:353]: done

File: Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py (hunk 2 above)
Comment (GuyAfik) [reply Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py:2]: removed it

File: Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py (hunk 4 above)
Comment (GuyAfik) [reply Packs/ServiceNow/Integrations/ServiceNowv2/test_data/response_constants.py:3011]: done

File: Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py (hunk 6 above)
Comment (GuyAfik) [reply Packs/ServiceNow/Integrations/ServiceNowv2/ServiceNowv2_test.py:360]: done

File: Packs/ServiceNow/Integrations/ServiceNowv2/README.md (hunk 7 above)
Comment (GuyAfik) [reply Packs/ServiceNow/Integrations/ServiceNowv2/README.md:113]: You are right, ill update the doc-string and ill update the description file as well

**Comment by GuyAfik** [comment]
@dantavori validation fails on that:
image

**Comment by ShahafBenYakir** [comment]
@GuyAfik first one will be fixed in SDK ([issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-415)) but I think the rest should be added to docs.

**Comment by GuyAfik** [comment]
> @GuyAfik first one will be fixed in SDK ([issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-415)) but I think the rest should be added to docs.

done with proper examples as well

**Review by dantavori — APPROVED** [review]

**Comment by dantavori** [comment]
fyi @yaakovpraisler

**Comment by GuyAfik** [comment]
@ShahafBenYakir @dantavori i think we are gonna need a force merge here because of this - 
image

**Pull request merged**